.github/workflows/test.yml

```diff
@@ -0,0 +1,344 @@
+# This is our main "CI tests" workflow. It runs everything that should run on
+# both PRs and merged commits, and for the latter reports failures to slack.
+name: CI
+
+on:
+  push:
+    branches:
+      - "main"
+      - "release-branch/*"
+  pull_request:
+    branches:
+      - "*"
+
+concurrency:
+  # For PRs, later CI runs preempt previous ones. e.g. a force push on a PR
+  # cancels running CI jobs and starts all new ones.
+  #
+  # For non-PR pushes, concurrency.group needs to be unique for every distinct
+  # CI run we want to have happen. Use run_id, which in practice means all
+  # non-PR CI runs will be allowed to run without preempting each other.
+  group: ${{ github.workflow }}-$${{ github.pull_request.number || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  test:
+    strategy:
+      fail-fast: false # don't abort the entire matrix if one element fails
+      matrix:
+        include:
+          - goarch: amd64
+          - goarch: amd64
+            variant: race
+          - goarch: "386" # thanks yaml
+    runs-on: ubuntu-latest
```

Comment:
We've been burned by "latest" already in the past. I'd rather hard-code a specific version. It should be our choice when the environment moves beneath us, not a surprise.

Reply:
Sure, we can do that. The counterpoint is that testing on -latest gives us some degree of alert that a new ubuntu broke something. But github updates after ubuntu releases, so by then it's too late anyway.

Reply:
Done. Pinned to ubuntu 22.04 and windows 2022, both the current "latest" per https://github.com/actions/runner-images